File: python/src/shares_v2.py
#!/usr/bin/env python3
# laurent.martin.aspera@fr.ibm.com
# Upload files Shares (Node) API and transfer spec v2
# Note: Transfer SDK may have a bug that make this work only if the share name is equal to the folder name on node.
import utils.configuration
import utils.transfer_client

config = utils.configuration.Configuration()
transfer_client = utils.transfer_client.TransferClient(config).startup()

try:
    shares_api_url = f'{config.param('shares', 'url')}/node_api'
    destination_folder = config.param('shares', 'folder_upload')
    # if '/' not in destination_folder:
    #    destination_folder = f'{config.param('shares','folder_upload']}/'
    # share_name, subfolder = destination_folder.split('/', 1)

    # prepare transfer spec v2 for COS
    t_spec = {
        'title': 'send using Node API and ts v2',
        'session_initiation': {
            'node_api': {
                'url': shares_api_url,
                'headers': [
                    tools.basic_auth_header_key_value(config.param('shares', 'username'), config.param('shares', 'password'))
                ]
            }
        },
        'direction': 'send',
        'assets': {
            'destination_root': destination_folder,
            'paths': []
        }
    }

    # add file list in transfer spec
    for f in config.file_list():
        # note: Shares API requires both source and destination to be set (unlike real node api)
        destination = f.split('/')[-1]
        # if subfolder:
        # destination = f'{destination_folder}/{destination}'
        t_spec['assets']['paths'].append(
            {'source': f,
             'destination': destination})

    # start transfer, using Transfer SDK
    transfer_client.start_transfer_and_wait(t_spec)
finally:
    transfer_client.shutdown()

File: cpp/src/shares_v2.cpp
#include "test_environment.hpp"
#undef LOGGER
#define LOGGER(level) BOOST_LOG_SEV(test_env.log(), boost::log::trivial::level)
int main(int argc, char* argv[]) {
    TestEnvironment test_env(argc, argv);
    const std::string shares_api_url = test_env.conf_str({"shares", "url"}) + "/node_api";
    LOGGER(info) << "Shares API URL: " << shares_api_url;
    // create V2 transfer spec
    json::object transfer_spec = json::object{
        {"title", "send using Node API and ts v2"},
        {"session_initiation", json::object{{"node_api",
                                             {{"url", shares_api_url},
                                              {
                                                  "headers",
                                                  json::array{{{"key", "Authorization"},
                                                               {"value", TestEnvironment::basic_auth_header(test_env.conf_str({"shares", "user"}), test_env.conf_str({"shares", "pass"}))}}},
                                              }}}}},
        {"direction", "send"},
        {"assets", json::object{{"destination_root", test_env.conf_str({"shares", "folder_upload"})},
                                {"paths", json::array()}}}};
    test_env.add_files_to_ts(transfer_spec["assets"].as_object()["paths"].as_array(), true);
    test_env.start_transfer_and_wait(transfer_spec);
    return 0;
}
